# Homework 1


Hồ Thanh Nhân - 21127122

---

[1]

## 1. 
(i) Not learning. There is no learning or adaptation involved. The coin classification system is created by the developers. Obtaining exact coin specifications from the U.S. Mint.

(ii) Supervised Learning. There is an algorithm which uses a large set of labeled coins to find out the boundaries for classifying coins.

(iii) Reinforcement Learning. Because the computer develops a strategy repeatly and adjusting its strategy at each move to lead to the losing.

Therefore, the answer is: [d] (i) Not learning, (ii) Supervised Learning, (iii) Reinforcement Learning.


## 2.
(i) Classifying numbers into primes and non-primes: This problem can be solved without using Machine Learning. It can be solved by using a simple algorithm which can determine whether a number is prime or not.

(ii) Detecting potential fraud in credit card charges: This problem is the best suited for Machine Learning. It can be solved by using Supervised Learning type of Machine Learning to train the historical transaction data, allowing it to detect the potential fraud in credit card charges.

(iii) Determining the time it would take a falling object to hit the ground: This problem can be solved without using Machine Learning. It can be solved by using classical physics formulas and mathematics for an accurate result.

(iv) Determining the optimal cycle for traffic lights in a busy intersection: This problem is best suited for Machine Learning. It can be solved by using Reinforcement type of Machine Learning to determine the optimal cycle for traffic lights by using real-time data.

Therefore, the answer is: [a] (ii) and (iv)

## 3.
A: represent the event of the first ball is black.

B: represent the event of the second ball is black.

Probability that the second ball is also black: P(B|A) = P = ?

Suppose: PA is P(A)

In [38]:
PA = 1/2 + 1/2 * 1/2
PA

0.75

Suppose: PBA is P(B ∩ A)

PBA = 1/2 (when selecting the bag with 2 black balls)

In [39]:
PBA = 1/2
P = PBA / PA
P

0.6666666666666666

Therefore, the answer is: [d] 2/3

## 4.
Suppose: PO is P(one green marble)
Suppose: PT is P(ten green marbles)

In [40]:
PO = 1 - 0.55
PO

0.44999999999999996

In [41]:
PT = PO ** 10
PT

0.0003405062891601559

Therefore, the answer is: $$[b] \space 3.405 * 10^{-4}$$

## 5.


In [42]:
N = 1000

P is the probability that (at least) one of the samples has ν = 0.

P1 is the probability that there is no sample has ν = 0.

P2 is the probability that ν = 0 in a sample.

We have: P = 1 - P1

In [43]:
P2 = 0.45 ** 10
P2

0.00034050628916015635

In [44]:
P1 = (1 - P2) ** N
P1

0.71136880215019

In [45]:
P = 1 - P1
P

0.28863119784980995

Therefore, the answer is: $$[c] \space 0.289$$

## 6.
[a] g returns 1 for all three points.

Number of target functions agreeing with hypothesis on all 3 points: 1 (111)

Number of target functions agreeing with hypothesis on exactly 2 points: 3 (011, 101, 110)

Number of target functions agreeing with hypothesis on exactly 1 point: 3 (001, 010, 100)

=> Score = 1 x 3 + 3 x 2 + 3 x 1 = 12

[b] g returns 0 for all three points.

Number of target functions agreeing with hypothesis on all 3 points: 1 (000)

Number of target functions agreeing with hypothesis on exactly 2 points: 3 (001, 010, 100)

Number of target functions agreeing with hypothesis on exactly 1 point: 3 (011, 101, 110)

=> Score = 1 x 3 + 3 x 2 + 3 x 1 = 12

[c] g is the XOR function applied to x, i.e., if the number of 1s in x is odd, g returns 1; if it is even, g returns 0.

We have: With the 3 points are 101, 110, 111, their target functions will be 0, 0, 1 respectively.

Number of target functions agreeing with hypothesis on all 3 points: 1 (001)

Number of target functions agreeing with hypothesis on exactly 2 points: 3 (000, 011, 101)

Number of target functions agreeing with hypothesis on exactly 1 point: 3 (010, 100, 111)

=> Score = 1 x 3 + 3 x 2 + 3 x 1 = 12

[d] g returns the opposite of the XOR function: if the number of 1s is odd, it returns 0, otherwise returns 1.

We have: With the 3 points are 101, 110, 111, their target functions will be 1, 1, 0 respectively.

Number of target functions agreeing with hypothesis on all 3 points: 1 (110)

Number of target functions agreeing with hypothesis on exactly 2 points: 3 (010, 100, 111)

Number of target functions agreeing with hypothesis on exactly 1 point: 3 (000, 011, 101)

=> Score = 1 x 3 + 3 x 2 + 3 x 1 = 12

Therefore, the answer is: [e] They are all equivalent (equal scores for g in [a] through [d]).

## 7-10. 

### Import neccessary libraries

In [46]:
import numpy as np

### Function definitions

Generating `target_w`, the vector of parameters of $f$ function

In [47]:
def generate_target_w():
    """
    Generates target_w from two random, uniformly distributed points in [-1, 1] x [-1, 1].
    
    Returns
    -------
    target_w : numpy array, shape (3, 1) 
        The vector of parameters of f.
    """
    # Generate two points from a uniform distribution over [-1, 1]x[-1, 1]
    p1 = np.random.uniform(-1, 1, 2)
    p2 = np.random.uniform(-1, 1, 2)
    # Compute the target W from these two points
    target_w = np.array([p1[1] * p2[0] - p1[0] * p2[1], p2[1] - p1[1], p1[0] - p2[0]]).reshape((-1, 1))
    
    return target_w

Generating data set function

In [48]:
def generate_data(N, target_w):
    """
    Generates a data set by generating random inputs and then using target_w to generate the 
    corresponding outputs.
    
    Parameters
    ----------
    N : int
        The number of examples.
    target_w : numpy array, shape (3, 1) 
        The vector of parameters of f.
    
    Returns
    -------
    X : numpy array, shape (N, 3)
        The matrix of input vectors (each row corresponds to an input vector); the first column of 
        this matrix is all ones.
    Y : numpy array, shape (N, 1)
        The vector of outputs.        
    """
    X = np.random.uniform(-1, 1, (N, 2))
    X = np.hstack((np.ones((N, 1)), X)) # Add 'ones' column
    Y = np.sign(np.dot(X, target_w))
    
    return X, Y

PLA function

In [49]:
def h(w, x):
    return np.sign(np.dot(x, w))

In [50]:
def has_converged(X, Y, w):
    return np.array_equal(h(w, X), Y)

In [51]:
def run_PLA(X, Y):
    """
    Runs PLA.
    
    Parameters
    ----------
    X : numpy array, shape (N, 3)
        The matrix of input vectors (each row corresponds to an input vector); the first column of 
        this matrix is all ones.
    Y : numpy array, shape (N, 1)
        The vector of outputs.
    
    Returns
    -------
    w : numpy array, shape (3, 1) 
        The vector of parameters of g.
    num_iterations : int
        The number of iterations PLA takes to converge.
    """
    w = np.zeros((X.shape[1], 1)) # Init w
    iteration = 0
    
    # TODO

    N = X.shape[0]

    while True:
        iteration += 1
        mix_id = np.random.permutation(N)
        for i in range(N):
            xi = X[mix_id[i], :]
            yi = Y[mix_id[i], 0]
            if h(w, xi)[0] != yi: # misclassified point
                w = w + (yi * xi).reshape(-1, 1)
                iteration += 1

        if has_converged(X, Y, w):
            break
        
    return w, iteration

# [2]

Main function

In [52]:
def main(N):
    """
    Parameters
    ----------
    N : int
        The number of training examples.
    """
    num_runs = 1000
    avg_num_iterations = 0.0 # The average number of iterations PLA takes to converge
    avg_test_err = 0.0 # The average test error of g - the final hypothesis picked by PLA
    
    for r in range(num_runs):
        # Generate target_w
        target_w = generate_target_w()
        
        # Generate training set
        X, Y = generate_data(N, target_w)
        
        # Run PLA to pick g
        w, num_iterations = run_PLA(X, Y)
        
        # Generate test set
        X_test, Y_test = generate_data(10000, target_w)
        
        # Test g
        test_err = np.mean(np.sign(np.dot(X_test, w)) != Y_test)
        
        # Update average values
        avg_num_iterations += (num_iterations * 1.0 / num_runs)
        avg_test_err += (test_err * 1.0 / num_runs)
    
    # Print results
    print('avg_num_iterations = %f' % (avg_num_iterations))
    print('avg_test_err = %f' % (avg_test_err))

### Run

In [53]:
main(N=10) # We can use `main(10)`, but `main(N=10)` is clearer

avg_num_iterations = 13.724000
avg_test_err = 0.107938


7. avg_num_iterations is closest to the answer [b] 15.

8. avg_test_err is closest to the answer [c] 0.1.

In [54]:
main(N=100)

avg_num_iterations = 134.225000
avg_test_err = 0.013310


9. avg_num_iterations is closest to the answer [b] 100

10. avg_test_err is closest to the answer [b] 0.01

References:

   [1] Github, Florian Peter, Caltech Machine Learning Homework # 1, last commit date: 27/04/2021, access date: 31/10/2023, https://github.com/workflow/caltech-machine-learning-homework/blob/master/HW1.ipynb

   [2] Tiep Vu Huu, Bài 9: Perceptron Learning Algorithm, Jan 21, 2017, access date: 31/10/2023, https://machinelearningcoban.com/2017/01/21/perceptron/